ocaml/typing/typecore.ml

```diff
@@ -6546,15 +6578,66 @@ and type_expect_
       end;
       let exp_extra = (Texp_stack, loc, []) :: exp.exp_extra in
       {exp with exp_extra}
-  | Pexp_overwrite (_x, _e) ->
+  | Pexp_overwrite (exp1, exp2) ->
       if not (Language_extension.is_enabled Overwriting) then
-        raise (Typetexp.Error (loc, env,
-                               Unsupported_extension Overwriting))
-      else Location.todo_overwrite_not_implemented loc
+        raise (Typetexp.Error (loc, env, Unsupported_extension Overwriting));
+      if not (can_be_overwritten exp2.pexp_desc) then
+        raise (Syntaxerr.(Error(Expecting(exp2.pexp_loc, "tuple, constructor or record"))));
+      let cell_mode =
+        let cell_mode = Value.newvar () in
+        (* The overwritten cell has to be unique
+           and should have the areality expected here: *)
+        Value.submode_exn cell_mode
+          (Value.max_with (Monadic Uniqueness) Uniqueness.unique);
+        Regionality.submode_exn
+          (Value.proj (Comonadic Areality) cell_mode)
+          (Value.proj (Comonadic Areality) expected_mode.mode);
+        cell_mode
+      in
+      let cell_type =
+        (* CR uniqueness: this could be the jkind of exp2 *)
+        { ty = newvar (Jkind.Builtin.value ~why:Boxed_record);
+          explanation = None }
+      in
+      let exp1 = type_expect ~recarg env (mode_default cell_mode) exp1 cell_type in
```

Comment:
As discussed in the meeting: this enables strong update. I can not use `type_exp` since I need to keep the type variable around to be passed down to the holes in the children. But I will use the `mk_expected` function to clean this up.